docs/internal/generics.md

```diff
@@ -156,3 +156,110 @@ let z : U32 & imm = obj.convert(y); // ERROR: Not able to define what type y is
+
+The compiler generates the required specialisations and replaces each definition with its respective concrete classes, then removes the generic class.
+Since generic classes can only be used by specialisation, all other (potentially infinite) possibilities of types parameters don't exist in the program, so the generic class is elided and we're left with only the specialisations.
+
+The end result would be:
+```ts
+// Definition of monomorphic classes only
+class Generic_U32 { field : U32 & imm; ... };
+class Generic_MyClass { field : MyClass & imm; ... }
+
+// Direct use of the concrete classes
+let foo : Generic_U32 = ...;
+let bar : Generic_MyClass = ...;
+```
+
+Note that type unions (`A | B`) are still monomorphic types.
+To resolve which implementation works with each inner type (`A` or `B`) one must use `match` inside the method.
+
+Example:
+```ts
+class Generic[A] {
+  field : A & imm;
+  // Implement constructor with match on types
+  add(other : A & imm) {
+    match (A)
+    { Numeric => field + other }
+    { String => field.append(other) }
+    { _ => Throw[None] } // Or some better error handling
+  }
+}
+...
+// This creates a monomorphic union type
+let foo : Generic[U32 | String] = ...;
+
+// This creates a monomorphic single type
+let bar : Generic[U32] = ...;
+
+// This union of Numeric types isn't quite static
+let baz : Generic[U32 | F64] = ...;
+```
+
+On the first case, `foo` is a monomorphic union between `U32` and `String`, which means, at run-time, either can be passed to `add` and we can't decide at compile time.
+So the implementation retains the `match` and will check if the types are `Numeric` or `String`.
+Note that both `field` and `other` must have the same type (both are `A`).
+
+However, on the second case, `bar` is a compile-time single monomorphic type, which guarantees that it will always implement the `Numeric` interface and therefore there's only one `match` statement that fits.
+In this case, the compiler is free to elide the `match` altogether and emit only the case where the type is `Numeric`, ie. `field + other`.
+In this case, the type is guaranteed to be `U32` to the call to `+` is a static call to `U32::+(U32, U32)`.
+
+Finally, in the last case, `baz`'s types are both `Numeric`, but the `+` operator for each is different.
+So, while the compiler is free to elide all other `match` cases, the dispatch of the `+` method is dynamic, ie. `field.+(other)`.
+
+The code above results in the following compile-time concrete types:
+```ts
+// (U32 | String) need different initialisation and add match
+class Generic_U32_String {
+  field : (U32 | String) & imm;
+  // Implement constructor with match on types
+  add(other : (U32 | String) & imm) {
+    match (typeof(field))
+    { U32 => U32::+(field, other) }
+    { String => String::append(field, other) }
```

Comment:
`String::append` might be defined to allow a `U32` to be appended, but it probably isn't what you wanted to do here.